File: python/83_path_sum_four_ways.py
import heapq as hq

def path_sum_four_ways(matrix: list) -> int:
    rows = len(matrix)
    cols = len(matrix[0])
    weights = [[float("inf")] * rows for _ in range(cols)]
    weights[0][0] = matrix[0][0]
    
    # Djikstra's algorithm
    pq = [(matrix[0][0], 0, 0)]
    dirs = [-1, 0, 1, 0, -1]
    
    while len(pq) > 0:
        cell = hq.heappop(pq)
        r = cell[1]
        c = cell[2]
        
        for i in range(4):
            new_row = r + dirs[i]
            new_col = c + dirs[i + 1]
            
            if new_row < 0 or new_row >= rows or new_col < 0 or new_col >= cols:
                continue
                
            new_weight = weights[r][c] + matrix[new_row][new_col] 
            
            if new_weight < weights[new_row][new_col]:
                weights[new_row][new_col] = new_weight
                hq.heappush(pq, (new_weight, new_row, new_col))
    
    return weights[-1][-1]
    

if __name__ == "__main__":
    f = open("83_matrix.txt", 'r')
    matrix = []
    for line in f.readlines():
        matrix.append([int(num) for num in line.strip().split(",")])
    f.close()
    
    # matrix = [
        # [131, 673, 234, 103, 18],
        # [201, 96, 342, 965, 150],
        # [630, 803, 746, 422, 111],
        # [537, 699, 497, 121, 956],
        # [805, 732, 524, 37, 331]
    # ]
    
    print(path_sum_four_ways(matrix))

File: cpp/83_path_sum_four_ways.cpp
#include <iostream>
#include <fstream>
#include <cstring>
#include <sstream>
#include <vector>
#include <queue>
#include <limits>

// minheap ordering of priority queue
struct ComparePath{
    bool operator()(const std::pair<int, std::pair<int, int>>& p1, const std::pair<int, std::pair<int, int>>& p2) {
        return p1.first > p2.first;
    }
};

int path_sum_four_ways(std::vector<std::vector<int>> matrix){
    int rows = matrix.size(), cols = matrix[0].size();
    // store minimum path sums to reach each cell
    std::vector<std::vector<int>> pathSums;
    for (int r = 0; r < rows; r++){
        std::vector<int> row;
        for (int c = 0; c < cols; c++){
            // initialize sum as maximum integer value
            row.push_back(std::numeric_limits<int>::max());
        }
        pathSums.push_back(row);
    }
    // starting at top left cell, ending at bottom right cell
    pathSums[0][0] = matrix[0][0];
    
    // store (path_sum, (r, c)) in priority queue, ordered by smallest path sum first
    std::priority_queue<std::pair<int, std::pair<int, int>>,
                        std::vector<std::pair<int, std::pair<int, int>>>,
                        ComparePath> pq;
    pq.push({pathSums[0][0], {0, 0}});
    
    while (!pq.empty()){
        // top does not pop from queue, pop does not return value
        int r = pq.top().second.first, c = pq.top().second.second;
        pq.pop();
        
        // check neighboring cells
        int dirs[] = {-1, 0, 1, 0, -1};
        for (int k = 0; k < 4; k++){
            int newR = r + dirs[k], newC = c + dirs[k + 1];
            if (0 <= newR and newR < rows and 0 <= newC and newC < cols){
                // continue path from neighboring cell if better path found
                int path = pathSums[r][c] + matrix[newR][newC];
                if (path < pathSums[newR][newC]){
                    pathSums[newR][newC] = path;
                    pq.push({pathSums[newR][newC], {newR, newC}});
                }
            }
        }
    }
    // return minimum path sum from top left to bottom right
    return pathSums[rows - 1][cols - 1];
}

int main(){
    std::ifstream file("/uploads/83_matrix.txt");
    std::string s;
    std::vector<std::vector<int>> matrix;
    
    // parse text file and extract matrix entries
    while (std::getline(file, s)){
        std::vector<int> row;
        std::stringstream ss(s);
        std::string num;
        while (std::getline(ss, num, ',')){
            row.push_back(std::stoi(num));
        }
        matrix.push_back(row);
    }
    file.close();
    
    std::cout << path_sum_four_ways(matrix) << "\n";
}
